components/object_store/src/obkv/mod.rs

```diff
@@ -0,0 +1,913 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    collections::{hash_map::DefaultHasher, HashSet},
+    hash::{Hash, Hasher},
+    ops::Range,
+    sync::{
+        atomic::{AtomicI64, AtomicU64, Ordering},
+        Arc,
+    },
+    time,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use chrono::{TimeZone, Utc};
+use common_util::error::{BoxError, GenericError};
+use futures::{
+    stream::{BoxStream, FuturesOrdered},
+    StreamExt,
+};
+use log::debug;
+use snafu::{ResultExt, Snafu};
+use table_kv::{ScanContext, ScanIter, TableKv, WriteBatch, WriteContext};
+use tokio::{
+    io::{AsyncWrite, AsyncWriteExt},
+    time::Instant,
+};
+use upstream::{
+    path::{Path, DELIMITER},
+    Error as StoreError, GetResult, ListResult, MultipartId, ObjectMeta, ObjectStore, Result,
+};
+
+use crate::{
+    multipart::{CloudMultiPartUpload, CloudMultiPartUploadImpl, UploadPart},
+    obkv::meta::{MetaManager, ObkvObjectMeta, META_TABLE},
+};
+
+mod meta;
+mod util;
+
+/// The object store type of obkv
+pub const OBKV: &str = "OBKV";
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to scan data, namespace:{}, err:{}", namespace, source,))]
+    ScanData {
+        namespace: String,
+        source: GenericError,
+    },
+
+    #[snafu(display("Failed to put data, path:{}, err:{}", path, source,))]
+    PutData {
+        namespace: String,
+        path: String,
+        source: GenericError,
+    },
+
+    #[snafu(display(
+        "Failed to create shard table, table_name:{}, err:{}",
+        table_name,
+        source,
+    ))]
+    CreateShardTable {
+        table_name: String,
+        source: GenericError,
+    },
+
+    #[snafu(display("Failed to read meta, path:{}, err:{}", path, source))]
+    ReadMeta { path: String, source: GenericError },
+
+    #[snafu(display("Empty data found in path:{}, index:{}", path, index))]
+    EmptyDataPart { path: String, index: usize },
+
+    #[snafu(display("No meta found, path:{}", path,))]
+    MetaNotExists { path: String },
+
+    #[snafu(display("Data is too large to put, size:{size}, limit:{limit}"))]
+    TooLargeData { size: usize, limit: usize },
+}
+
+impl<T: TableKv> MetaManager<T> {
+    fn try_new(client: Arc<T>) -> std::result::Result<Self, Error> {
+        create_table_if_not_exists(client.clone(), META_TABLE)?;
+        Ok(Self { client })
+    }
+}
+
+/// If table not exists, create shard table; Else, do nothing.
+fn create_table_if_not_exists<T: TableKv>(
```

Comment:
Why not define this method on `ShardManager`?

Reply:
This method is not only used for shard table, also used for meta table.